### ISSUE TITLE
[BUG] ODrive corrupted my file on Googledrive

### ISSUE THREAD

**Issue by nophead** [liberodark/ODrive#134]
**Describe the bug**
ODrive replaced the contents of one of my files with a JSON exception and wrote it back to GoogleDrive without changing the date so I have no idea when it did this.

**To Reproduce**
No idea.

**Expected behavior**
File contents should not be corrupted.

**Screenshots**
```
{
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "authError",
    "message": "Invalid Credentials",
    "locationType": "header",
    "location": "Authorization"
   }
  ],
  "code": 401,
  "message": "Invalid Credentials"
 }
}
```
**Desktop**
 - OS: [Ubuntu 18.04]
 - Version [@HEAD with my own rate limit patch.]

**Additional context**
The file happened to be part of an XML database so the program that parsed it threw an error.  I have no idea if any of my tens of thousands of other files has been trashed.

**Comment by nophead** [comment]
The name of the corrupted file is Assert.palmerc, so I don't know if that was an ironic coincidence or somehow created the bug.

**Comment by nophead** [comment]
Another file simply has this in it:
```
Service Unavailable
```

**Comment by nophead** [comment]
Comparing 21,195 Files in 2,604 Folders with a backup I made before trying ODRIVE it seems that only the two files I mentioned are corrupt.

**Comment by yuhc** [comment]
Thanks for sharing @nophead .
This is really a severe bug and makes me pause using odrive.
I'd wait and look forward to the analysis and fix of this issue.

**Comment by nophead** [comment]
Yes I have had to stop using it as it is the worst possible bug. It must not be checking a return status at some point and takes the error message as the file contents.

**Comment by liberodark** [comment]
Better to user the 0.2.1 version is more stable

**Comment by pmarrapese** [comment]
I ran into this too on 0.3.0.

```
{
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "internalError",
    "message": "Internal Error"
   }
  ],
  "code": 500,
  "message": "Internal Error"
 }
}
```

**Comment by pmarrapese** [comment]
This is happening on 0.2.1 as well...

```
{
 "error": {
  "errors": [
   {
    "domain": "usageLimits",
    "reason": "userRateLimitExceeded",
    "message": "User Rate Limit Exceeded. Rate of requests for user exceed configured project quota. You may consider re-evaluating expected per-user traffic to the API and adjust project quota limits accordingly. You may monitor aggregate quota usage and adjust limits in the API Console: https://console.developers.google.com/apis/api/drive.googleapis.com/quotas?project=985525764653",
    "extendedHelp": "https://console.developers.google.com/apis/api/drive.googleapis.com/quotas?project=985525764653"
   }
  ],
  "code": 403,
  "message": "User Rate Limit Exceeded. Rate of requests for user exceed configured project quota. You may consider re-evaluating expected per-user traffic to the API and adjust project quota limits accordingly. You may monitor aggregate quota usage and adjust limits in the API Console: https://console.developers.google.com/apis/api/drive.googleapis.com/quotas?project=985525764653"
 }
}
```

**Comment by nophead** [comment]
Yes I couldn't use the 0.2.1 version because of the rate limit problem. As this bug states I used the latest HEAD with my own path for the rate limit bug that I detail here: https://github.com/liberodark/ODrive/issues/99#issuecomment-525251036

I presume this bug is caused by not checking an error return value somewhere.

**Comment by liberodark** [comment]
Yeah is a API limit.
Google make a limite from the API for download files / make request.
Is a big problem.
One google not have native client and two is make a limit for projects...

**Comment by cserpell** [comment]
Does Google Drive keep a copy of the old file in history when browsing through web client?

**Comment by nophead** [comment]
Where would I find a file history?

The problem is I have 20000 files on Google drive, how would I know when one gets randomly corrupted? I only noticed because it was an part of a database and the reader told me it was corrupt. Fortunately I had a backup.

**Comment by cserpell** [comment]
I get your point. Just for the record, when browsing your files, click over one and at the right hand side, there is an "Activity" panel with versions.

**Comment by mihafiz** [comment]
I'm not using odrive because of this. Critically serious bug still open and unresolved!

**Comment by nophead** [comment]
I moved on to InSync. Not free but a one off payment and seems on a par with Google's backup and sync.
